tsdb/record/record_test.go

```diff
@@ -160,3 +184,71 @@ func TestRecord_Type(t *testing.T) {
 	recordType = dec.Type([]byte{0})
 	require.Equal(t, Unknown, recordType)
 }
+
+func TestRecord_MetadataDecodeUnknownFields(t *testing.T) {
+	var enc encoding.Encbuf
+	var encFields encoding.Encbuf
+	var dec Decoder
+
+	// Write record type.
+	enc.PutByte(byte(Metadata))
+
+	// Write first metadata entry, all known fields.
+	enc.PutUvarint64(101)
+	encFields.Reset()
+	encFields.PutByte(byte(Counter))
+	encFields.PutUvarintStr("")
+	encFields.PutUvarintStr("some magic counter")
+	// Write size of encoded fields then the encoded fields.
+	enc.PutUvarint(encFields.Len())
+	enc.B = append(enc.B, encFields.B...)
+
+	// Write second metadata entry, known fields + unknown fields.
+	enc.PutUvarint64(99)
+	encFields.Reset()
+	// Known fields.
+	encFields.PutByte(byte(Counter))
+	encFields.PutUvarintStr("seconds")
+	encFields.PutUvarintStr("CPU time counter")
+	// Unknown fields.
+	encFields.PutUvarintStr("some unknown field")
+	encFields.PutByte('a')
```

Comment:
Is this intentionally reversed? Also there was no other place in the code which hinted that there could be more fields. Is it documented somewhere?

Reply:
This specific testcase also ties around the general `size` field idea and the question on using a manual encoder in `TestRecord_MetadataDecodeUnknownFields`.

These three lines are just adding some garbage fields that we currently have no knowledge of, just to check that when we're decoding metadata later on, they're skipped over instead of tripping the decoder into producing an error.
```
	// Unknown fields.
	encFields.PutUvarintStr("some unknown field")
	encFields.PutByte('a')
	encFields.PutUvarint64(42)
```